apis/metrics/v1alpha1/metricsconfiguration_types.go

```diff
@@ -0,0 +1,312 @@
+	metav1.TypeMeta   `json:",inline,omitempty"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Spec              MetricsConfigurationSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// MetricsConfigurationSpec is the spec of MetricsConfiguration object.
+type MetricsConfigurationSpec struct {
+	// TargetRef defines the object for which metrics will be collected
+	TargetRef TargetRef `json:"targetRef" protobuf:"bytes,1,opt,name=targetRef"`
+
+	// List of Metrics configuration for the resource object defined in TargetRef
+	Metrics []Metrics `json:"metrics" protobuf:"bytes,2,rep,name=metrics"`
+}
+
+// TargetRef contains the Object's group and resource to specify the target resource
+type TargetRef struct {
+	// Group defines the group of the object.
+	// Example: For Deployment, Group will be 'apps'
+	Group string `json:"group" protobuf:"bytes,1,opt,name=group"`
+
+	// Resource defines the resource of the object.
+	// Example: For Deployment, Resource will be 'deployments'
+	Resource string `json:"resource" protobuf:"bytes,2,opt,name=resource"`
+}
+
+// Metrics contains the configuration of a metric in prometheus style.
+type Metrics struct {
+	// Name defines the metrics name. Name should be in snake case.
+	// Example: kube_deployment_spec_replicas
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// Help is used to describe the metrics.
+	// Example: For kube_deployment_spec_replicas, help string can be "Number of desired pods for a deployment."
+	Help string `json:"help" protobuf:"bytes,2,opt,name=help"`
+
+	// Type defines the metrics type.
+	// For kubernetes based object, types can only be gauge
+	Type string `json:"type" protobuf:"bytes,3,opt,name=type"`
+
+	// Field defines the metric value path of the manifest file and the type of that value
+	// +optional
+	Field Field `json:"field,omitempty" protobuf:"bytes,4,opt,name=field"`
+
+	// Labels defines the metric labels as a key-value pair
+	// +optional
+	Labels []Label `json:"labels,omitempty" protobuf:"bytes,5,rep,name=labels"`
+
+	// Params is list of parameters configuration used in expression evaluation
+	// +optional
+	Params []Parameter `json:"params,omitempty" protobuf:"bytes,6,rep,name=params"`
+
+	// States handle metrics with label cardinality.
+	// States specify the possible states for a label
+	// and their corresponding MetricValue configuration.
+	//
+	// Metrics must contain either States or MetricValue.
+	// If both are specified, MetricValue will be ignored.
+	// +optional
+	States State `json:"states,omitempty" protobuf:"bytes,7,opt,name=states"`
+
+	// MetricValue defines the configuration to obtain metric value.
+	//
+	// Metrics must contain either States or MetricValue.
+	// If both are specified, MetricValue will be ignored.
+	// +optional
+	MetricValue MetricValue `json:"metricValue,omitempty" protobuf:"bytes,8,opt,name=metricValue"`
+}
+
+type FieldType string
+
```

Comment:
Add 

`// +kubebuilder:validation:Enum=Integer;DateTime;Array;String`

Reply:
Added and re-generated crd.